File: backend/app/entrypoints/api/routes/planner.py
"""Spaced-repetition planner routes."""

from __future__ import annotations

from fastapi import APIRouter, Query

from app.application.planner_use_case import PlannerUseCase
from app.entrypoints.api.deps import ContainerDep, CurrentUserDep
from app.entrypoints.api.schemas import ReviewRequest

router = APIRouter(prefix="/planner", tags=["planner"])


def _use_case(container: ContainerDep) -> PlannerUseCase:
    return PlannerUseCase(subjects=container.subjects, study=container.study)


@router.get("/due")
async def list_due(
    user: CurrentUserDep,
    container: ContainerDep,
    subject_id: str = Query(...),
    limit: int = Query(default=20, ge=1, le=50),
) -> dict[str, object]:
    items = await _use_case(container).due(user_id=user.id, subject_id=subject_id, limit=limit)
    return {"items": items}


@router.post("/review")
async def review_card(
    body: ReviewRequest,
    user: CurrentUserDep,
    container: ContainerDep,
) -> dict[str, object]:
    return await _use_case(container).review(
        user_id=user.id,
        flashcard_id=body.flashcard_id,
        quality=body.quality,
    )

File: frontend/src/lib/api/planner.ts
import { apiRequest } from "@/lib/api/http";

export type DueCardDto = {
  flashcard_id: string;
  front: string;
  back: string;
  hint?: string | null;
  artifact_title: string;
  next_review_at?: string | null;
  repetitions: number;
};

export async function listDueCards(token: string, subjectId: string) {
  return apiRequest<{ items: DueCardDto[] }>("/planner/due", {
    token,
    query: { subject_id: subjectId },
  });
}

export async function reviewCard(token: string, flashcardId: string, quality: number) {
  return apiRequest<{
    flashcard_id: string;
    ease: number;
    interval_days: number;
    repetitions: number;
    next_review_at?: string | null;
  }>("/planner/review", {
    method: "POST",
    token,
    json: { flashcard_id: flashcardId, quality },
  });
}
